lib/ansible/module_utils/storage/hpe3par/hpe3par.py

```diff
@@ -0,0 +1,91 @@
+# Copyright: (c) 2018, Hewlett Packard Enterprise Development LP
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from ansible.module_utils import basic
+
+
+def convert_to_binary_multiple(size_with_unit):
+    if size_with_unit is None:
+        return -1
+    valid_units = ['MiB', 'GiB', 'TiB']
+    valid_unit = False
+    for unit in valid_units:
+        if size_with_unit.strip().endswith(unit):
+            valid_unit = True
+            size = size_with_unit.split(unit)[0]
+            if float(size) < 0:
+                return -1
+    if not valid_unit:
+        raise ValueError("%s does not have a valid unit. The unit must be one of %s" % (size_with_unit, valid_units))
+
+    size = size_with_unit.replace(" ", "").split('iB')[0]
+    size_kib = basic.human_to_bytes(size)
+    return int(size_kib / (1024 * 1024))
+
+
+storage_system_spec = {
+    "storage_system_ip": {
+        "required": True,
+        "type": "str"
+    },
+    "storage_system_username": {
+        "required": True,
+        "type": "str",
+        "no_log": True
+    },
+    "storage_system_password": {
+        "required": True,
+        "type": "str",
+        "no_log": True
+    },
+    "secure": {
+        "type": "bool",
+        "default": False
+    }
+}
+
+
+def cpg_argument_spec():
+    spec = {
+        "state": {
+            "required": True,
+            "choices": ['present', 'absent'],
+            "type": 'str'
+        },
+        "cpg_name": {
+            "required": True,
+            "type": "str"
+        },
+        "domain": {
+            "type": "str"
+        },
+        "growth_increment": {
+            "type": "str",
+        },
+        "growth_limit": {
+            "type": "str",
+        },
+        "growth_warning": {
+            "type": "str",
+        },
+        "raid_type": {
+            "required": False,
+            "type": "str",
+            "choices": ['R0', 'R1', 'R5', 'R6']
+        },
+        "set_size": {
+            "required": False,
+            "type": "int",
+            "default": -1
```

Comment:
It seems `set_size` should be required (then without a default value) since [the default value isn't a valid default value](https://github.com/HewlettPackard/hpe3par_python_sdk/blob/c4e508e6d11ff0fc2d173558d3497a2cb81d0f3b/hpe3par_sdk/client.py#L51).

Reply:
As per the [HPE 3PAR Web Sevices API document](https://support.hpe.com/hpsc/doc/public/display?docId=c03606339) - page 84, any negative value for set size is ignored. In that case, if the user does not specify a value for the set size, it is just ignored.

![image](https://user-images.githubusercontent.com/22393330/48303088-ebf73000-e52b-11e8-8a09-8cbc08234fb9.png)

In this case, if the value of set size is not provided, it is defaulted to -1 (a negative value) and is safely ignored.

Reply:
- either when `raid_type` isn't set by user (then `raid_type` value is `None`)
- or when `raid_type` is set by user but `set_size` isn't 

module returns an error and nothing is sent to the HPE 3PAR web service:
https://github.com/ansible/ansible/blob/053ff9998ac44a0ac7392591f1e8261e912351ec/lib/ansible/modules/storage/hpe3par/cpg_3par.py#L130-L135

- either `required_together` must be used ([example](https://github.com/ansible/ansible/blob/053ff9998ac44a0ac7392591f1e8261e912351ec/lib/ansible/modules/crypto/openssl_privatekey.py#L268))
- or `validate_set_size` check should be removed